### PR TITLE
WIP: Initial conversion of the ETL to prefect flow

### PR DESCRIPTION
**Pull request by rousik**
This is still very much WIP and is for informative purposes only so that you can familiarize yourself with the changes that this introduces.

Couple of important features:
1. adds support for uploading csv files, datapackages and parquet files to GCS
2. ferc1_to_sqlite is now executed as part of ETL
3. when datapackages are created, pudl database is populated with their contents
4. validation step is optional but it doesn't seem to work very well yet
5. epacems is not written to datapackages anymore but instead written directly to parquet files
6. Excel extractor has been turned into prefect task so it can be directly used to fetch data as a prefect task

Overall, the flow is constructed in `etl.py` using `${Dataset}Pipeline` objects that are responsible for adding relevant tasks to the flow. Many of the extract/transform methods are annotated with `@task` decorators which turns them into prefect tasks, sometimes simple wrappers are used to aggregate some steps (e.g. epacems).

`DataFrameCollection` is used as a default data structure for passing dataframes between prefect tasks. This will ensure that data is written to disk and not held in memory.

### CODE REVIEW

**Comment by rousik** [comment]
Here's list of tasks that should be addressed as part of this PR that came out of our walkthrough today:

- [ ] Prevent accidental collisions in DFC (`store()` should thrown an exception if the table already exists or if the file exists on disk)
- [ ] Datastore itself should be in charge of populating prefect.context and perhaps declaring commandline flags it needs to keep the logic contained in a single module.
- [ ] Move DatasetPipeline instances into `src/pudl/pipeline/${dataset}.py` to declutter `etl.py`
- [ ] Attempt to run validation by calling `pytest.main` with the necessary arguments (this should by pass `ShellTask` and `tox` which feel like unnecessary complications.

**Comment by rousik** [comment]
Moved `DatasetPipeline` implementations under `workflow/` directory (instead of originally proposed pipeline that seems to be already a rather overloaded term).

**Comment by rousik** [comment]
Validation tests are not distributed with the pudl package and hence it won't be easy to execute those in the docker etl container unless we make some more substantial changes. I think it's best to split solving that issue into its own issue and its own PR.

**Comment by rousik** [comment]
`Datastore` caching logic simplified (with fsspec there's no need for diverse cache implementations) and some code cleanups done. Now re-running the etl tests to see if this latest instance is still operating properly.

**Comment by nickrobinson251** [comment]
> with fsspec there's no need for diverse cache implementations

Tried with S3 buckets for both `--zenodo-cache-path` and `--upload-to` and it seems to work perfectly. Supersedes #917 / https://github.com/nickrobinson251/pudl/pull/2 

Thanks a lot! 

The `metafs` implementation is very clever 👍

**Comment by rousik** [comment]
Yeah "clever" is a dangerous word in the software development because it
implies complex magic which is happening here. But I hope the degree of
convenience makes it acceptable.

**Comment by nickrobinson251** [comment]
Yes, i'm a bit surprised fsspec doesn't itself provide an interface like `metafs`. I wonder if it's worth opening a feature request there?

> "clever" is a dangerous word

For sure. But I think the ability to support any fsspec-compatible path seems like a big win, and fortunately it's very few lines of code for the "shim" (and saves implementing/maintaining caches). I think it's nicely done! Thanks again for supporting the S3 use-case.

**Comment by bendnorman** [comment]
Ok! I’ve got dev merged into this branch with all the tests passing. I have also cleaned up the branch and tackled some TODOs Jan left. This branch is not complete ([see remaining issues](https://github.com/catalyst-cooperative/pudl/issues/840)) but I think the largest changes have occurred and I would love some feedback. I left some questions in my review.

File: src/pudl/cli.py [hunk 1]
```diff
@@ -26,24 +31,21 @@
...
-    Args:
-        argv (list): command line arguments including caller file name.
-
-    Returns:
-        dict: A dictionary mapping command line arguments to their values.
+def create_arg_parser():
+    """Create an argument parser for the etl script."""
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        parents=[pudl.etl.command_line_flags()])
```
Comment (bendnorman) [inline src/pudl/cli.py:38]: Jan split command args into etl.py and cli.py. Should all of them live in cli.py?

File: src/pudl/etl.py [hunk 2]
```diff
@@ -349,61 +210,100 @@ def etl(  # noqa: C901
...
-    ds_kwargs = dict(
-        gcs_cache_path=gcs_cache_path,
-        sandbox=pudl_settings.get("sandbox", False)
-    )
-    if use_local_cache:
-        ds_kwargs["local_cache_path"] = Path(pudl_settings["pudl_in"]) / "data"
-
-    validated_etl_settings = etl_settings.datasets
+    # Setup pipeline cache
+    configure_prefect_context(etl_settings, pudl_settings, commandline_args)
```
Comment (bendnorman) [inline src/pudl/etl.py:226]: Prefect setup is kind of split between cli and etl. I think I’ve been starting at this for too long and would love any ideas on how to make the setup cleaner.

File: src/pudl/etl.py [hunk 3]
```diff
@@ -15,326 +15,187 @@
...
+    if (commandline_args.keep_cache == 'NEVER') or onfail_success:
+        if not cache_root.startswith("gs://"):
+            logger.warning(f'Deleting pipeline cache directory under {cache_root}')
+            # TODO(rousik): in order to prevent catastrophic results due to misconfiguration
+            # we should refuse to delete cache_root unless the directory has the expected
+            # run_id form of YYYY-MM-DD-HHMM-uuid4
+            shutil.rmtree(cache_root)
+
+
+def configure_prefect_context(etl_settings, pudl_settings, commandline_args):
```
Comment (bendnorman) [inline src/pudl/etl.py:171]: Prefect.context is a global var store that can be accessed during a flow run. Things like etl_settings, pudl_settings, overwrite_ferc1_db and pipeline_cache_path are all stored in prefect.context. Do y’all have any opinions on what should and should not live in this?

File: src/pudl/workflow/epacems.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        """Add epacems tasks to the flow."""
+        with self.flow as flow:
+            plants = pudl.transform.epacems.load_plant_utc_offset()
+
+            # Wait to build CEMS until EIA is done if EIA is in the settings file.
+            # If EIA is not in the settings file, go ahead and build CEMS on its own.
+            if "eia" in prefect.context.get("datasets", {}).keys():
+                dfs_to_sqlite_task = flow.get_tasks(name="dfs_to_sqlite")
+                logger.info("Setting EIA as a dependency of CEMS.")
+                plants.set_dependencies(upstream_tasks=[dfs_to_sqlite_task])
```
Comment (bendnorman) [inline src/pudl/workflow/epacems.py:62]: Setting the eia and epacems dependency is a little awkward right now. EpaCemsPipeline has to look into the flow to see if the EIA pipeline is being run.

File: src/pudl/etl.py [hunk 5]
```diff
@@ -15,326 +15,187 @@
...
+        type=str,
+        default=os.environ.get('PUDL_UPLOAD_TO'),
+        help="""A location (local or remote) where the results of the ETL run
+        should be uploaded to. This path will be interpreted by fsspec so
+        anything supported by that module is a valid destination.
+        This should work with GCS and S3 remote destinations.
+        Default value for this will be loaded from PUDL_UPLOAD_TO environment
+        variable.
+        Files will be stored under {upload_to}/{run_id} to avoid conflicts.
+        """)
```
Comment (bendnorman) [inline src/pudl/etl.py:70]: Do we want to support upload_to right now?

File: src/pudl/cli.py (hunk 1 above)
Comment (cmgosnell) [reply src/pudl/cli.py:38]: Hm. Is there any practical reason for having the chunk of args in `etl.py`? On the surface it seems much cleaner to pull all of the command args over here.

File: test/unit/dfc_test.py [hunk 6]
```diff
@@ -1,114 +1,120 @@
...
 
         # Check that df added by reference points to the same underlying frame on disk.
-        self.dfc.add_reference("test", refs["test"])
-        self.assertEqual(["test"], self.dfc.get_table_names())
-        pd.testing.assert_frame_equal(test_df, self.dfc.get("test"))
-        self.assertEqual(self.dfc.references(), extra_dfc.references())
+        dfc_fixture.add_reference("test", refs["test"])
+        dfc_fixture.get_table_names() == ["test"]
+        pd.testing.assert_frame_equal(test_df, dfc_fixture.get("test"))
+        extra_dfc.references() == dfc_fixture.references()
```
Comment (zaneselvans) [inline test/unit/dfc_test.py:92]: Aren't assertions required here to check for equality?

File: test/unit/dfc_test.py [hunk 7]
```diff
@@ -117,59 +123,61 @@ def test_union_of_two(self):
         dfc2.store("second", second)
 
         final_dfc = dfc1.union(dfc2)
-        self.assertEqual(["first"], dfc1.get_table_names())
-        self.assertEqual(["second"], dfc2.get_table_names())
-        self.assertEqual(["first", "second"], final_dfc.get_table_names())
+        assert dfc1.get_table_names() == ["first"]
+        assert dfc2.get_table_names() == ["second"]
+        assert final_dfc.get_table_names() == ["first", "second"]
```
Comment (zaneselvans) [inline test/unit/dfc_test.py:128]: Is this particular ordering of names guaranteed? Would `["second", "first"]` indicate failure? Could convert both sides to sets.

File: test/unit/dfc_test.py [hunk 8]
```diff
@@ -117,59 +123,61 @@ def test_union_of_two(self):
...
 
         dfc1.store("x", pd.DataFrame())
         dfc1.store("y", pd.DataFrame())
         dfc2.store("y", pd.DataFrame())
         dfc2.store("z", pd.DataFrame())
-        self.assertRaises(dfc.TableExistsError, dfc1.union, dfc2)
 
-    def test_to_dict(self):
+        with pytest.raises(dfc.TableExistsError):
+            dfc1.union(dfc1)
```
Comment (zaneselvans) [inline test/unit/dfc_test.py:144]: Is this supposed to be unioning `dfc1` with itself? Or with `dfc2`?

File: test/unit/dfc_test.py [hunk 9]
```diff
@@ -193,7 +201,7 @@ def named_empty_df(df_name):
             names = ["a", "b", "c", "d"]
             res = dfc.merge_list(named_empty_df.map(names))
         final = f.run().result[res].result
-        self.assertEqual(["a", "b", "c", "d"], final.get_table_names())
+        assert final.get_table_names() == ["a", "b", "c", "d"]
```
Comment (zaneselvans) [inline test/unit/dfc_test.py:204]: Is the ordering guaranteed?

File: test/unit/dfc_test.py [hunk 10]
```diff
@@ -193,7 +201,7 @@ def named_empty_df(df_name):
...
 
     def test_fanout_task(self):
```
Comment (zaneselvans) [inline test/unit/dfc_test.py:206]: I'm bewildered by this test. I guess I need to go look at the DFC class. Do you understand what fanout is doing, and the work that Prefect is doing in here?

File: test/unit/settings_test.py [hunk 11]
```diff
@@ -211,3 +211,18 @@ def test_missing_dataset(self):
         """Test a KeyError is raised when a use requests a unavailable dataset."""
         with pytest.raises(KeyError):
             create_dataset_settings("oops!")
+
+
+class TestEtlSettings:
+    """Test EtlSettings."""
+
+    # TODO(bendnorman): The behavior of this method works and passes as an invidual unit test but fails when run with all unit tests.
```
Comment (zaneselvans) [inline test/unit/settings_test.py:219]: Weird. Is it some kind of issue with the order in which tests get run?

**Review by zaneselvans — CHANGES_REQUESTED** [review]
omg I got through the whole thing. Lots of comments and questions inline and I may bring some comments over from the Prefect Docs and put them in the comments on the PR.

File: src/pudl/etl.py (hunk 5 above)
Comment (bendnorman) [reply src/pudl/etl.py:70]: That is true. If we aren't running this remotely soon should it be included in this PR? I don't want to add functionality that isn't fully supported.

File: src/pudl/cli.py [hunk 12]
```diff
@@ -131,12 +237,7 @@ def main():
     pudl.etl.etl(
         etl_settings=etl_settings,
         pudl_settings=pudl_settings,
-        clobber=args.clobber,
-        use_local_cache=not args.bypass_local_cache,
-        gcs_cache_path=args.gcs_cache_path,
-        check_foreign_keys=not args.ignore_foreign_key_constraints,
-        check_types=not args.ignore_type_constraints,
-        check_values=not args.ignore_value_constraints,
```
[the remark this answers is not in the document]
Comment (bendnorman) [reply src/pudl/cli.py:139]: They are getting passed in through the command line args.

File: test/unit/dfc_test.py [hunk 13]
```diff
@@ -1,114 +1,120 @@
...
 
-    def test_empty_dfc(self):
+    def test_empty_dfc(self, dfc_fixture):
         """Tests that new DataFrameCollection is devoid of content."""
-        self.assertEqual([], self.dfc.get_table_names())
-        self.assertEqual([], list(self.dfc.references()))
-        self.assertEqual([], list(self.dfc.items()))
+        dfc_fixture.get_table_names() == []
+        list(dfc_fixture.references()) == []
+        list(dfc_fixture.items()) == []
```
[the remark this answers is not in the document]
Comment (bendnorman) [reply test/unit/dfc_test.py:35]: Ah yes. I am converting Jan's unittests to pytests and these ones slipped through the cracks.

**Comment by zaneselvans** [comment]
* Persistence of intermediate results seems useful for development/debugging, for data provenance, and for building other derived outputs (e.g. the "final" PUDL DB is itself also an intermediate result, since we then do analysis based on that data).
* Using FSSpec to make interacting with local & remote filesystems seems useful and simpler than having totally separate systems.
* Using our own system to persist results seems like it might not be ideal in the long run. What's the canonical way of doing this? My impression is that it's database tables. Feels like we need to talk about whether / how we are going to rely on having Postgres / BigQuery be a requirement for running the ETL, and more broadly our expectations about who is going to run the ETL and under what circumstances.
* I have a gut feeling that the way we're using Prefect here is kind of off-label, and might be annoying to maintain. It seems like there are a lot of explicitly constructed dependencies that are kind of hard to follow, rather than a high level coordinating module which composes all the tasks together "functionally" and lets you follow the flow of data as a narrative. Is this really necessary? Or is it an artifact of trying to preserve the existing structure of the codebase?
* If we do some thinking about which parts of the pipeline should really be getting run when... does the architecture change? If we keep the datasets separate from each other for longer?
* I wasn't totally clear on whether `ferc1_to_sqlite` and `epacems_to_parquet` were still available to be run independently.
* Do we have a sense of what doing development and testing with all of this being obligated to run on cloud resources would look like?
* Moving to using Zenodo as cold-storage that's mirrored to cloud storage initially seems like it would be great, especially if/when we go to doing continuous updates to the raw data archives, since that will mean more frequent downloads. Seems like splitting the Datastore stuff off as n independent package might make sense?
* Seems like as the codebase is set up right now, it'll be hard to only re-run parts of the pipeline when they change -- we don't really have any way of telling which parts have changed and will need an update.

**Comment by rousik** [comment]
> * Using our own system to persist results seems like it might not be ideal in the long run. What's the canonical way of doing this? My impression is that it's database tables. Feels like we need to talk about whether / how we are going to rely on having Postgres / BigQuery be a requirement for running the ETL, and more broadly our expectations about who is going to run the ETL and under what circumstances.

I suppose you're referring to the dataframe serialization to disk? I do not think there's any canonical/universal way of doing this and the implementation choice depends on your specific usecase.

This format essentially adds disk serialization feature to the preexisting data exchange format between tasks (collections of named dataframes) which allows you to 1. distribute the workload across machines, 2. reduces the memory pressure by offloading the intermediate states to disk.

That said, this is totally something that can be switched/replaced with a different serialization mechanism if needed/wanted. I would view this as internal/development format so we should be using whatever is useful to us w.r.t. debuggability/maintenance and integration with the pipeline internals.

In particular, we could use database but I think it would make sense only if:
1. db would provide superior performance and we consider performance to be critical
2. the pipeline itself (individual tasks) would operate on the database directly rather than processing dataframes (otherwise, we will likely have to do dataframe -> serialize to database -> extract dataframe from database -> process dataframe -> repeat)

> * I have a gut feeling that the way we're using Prefect here is kind of off-label, and might be annoying to maintain. It seems like there are a lot of explicitly constructed dependencies that are kind of hard to follow, rather than a high level coordinating module which composes all the tasks together "functionally" and lets you follow the flow of data as a narrative. Is this really necessary? Or is it an artifact of trying to preserve the existing structure of the codebase?

Can you expand on what you mean by off-label use of prefect?

The general idea is that `DatasetPipeline` should ultimately offer a high-level abstraction that allows end-user to construct the ETL for a given dataset and should be data-flow oriented. Ideally, we would be operating at the level of table rather than dataset which would be ideal but this would require further refactoring of the code to make it happen so some of the clunkiness may be related to preservation of the current code structure (with dataset granularity of tasks).

Overall I do think that it might be valuable to offer higher-level API where end-users/developers will describe how individual tables are constructed or transformed and then let some translation layer turn this into prefect tasks. This might have the benefit of insulating end-users from the lower-level abstractions (prefect tasks) and enforcing some common structure. I do think that the prototype is somewhat leaky as we are still dealing with a bit of both (DatasetPipeline combined with prefect task annotated functions).

> * If we do some thinking about which parts of the pipeline should really be getting run when... does the architecture change? If we keep the datasets separate from each other for longer?

Can you elaborate on this thought a bit more?

> * Do we have a sense of what doing development and testing with all of this being obligated to run on cloud resources would look like?

I think it should still be possible to run this locally so using cloud resources should not be mandatory. However, I do think that this would allow us to run some automated tests/validation on the cloud, e.g. as part of the PR validation (e.g. PR can kick of github action that will run the pipeline and emit the results of the pipeline somewhere so that they can be checked for correctness or compared against main branch to highlight if the results have changed or not (noop validation)).

> * Moving to using Zenodo as cold-storage that's mirrored to cloud storage initially seems like it would be great, especially if/when we go to doing continuous updates to the raw data archives, since that will mean more frequent downloads. Seems like splitting the Datastore stuff off as n independent package might make sense?

Yeah. I can even imagine situation where ETL would rely solely on the cloud storage mirror and another component/pipeline (datastore) would take care of synchronization. 

> * Seems like as the codebase is set up right now, it'll be hard to only re-run parts of the pipeline when they change -- we don't really have any way of telling which parts have changed and will need an update.

Automatically determining what needs to be rerun is a hard problem (perhaps practically intractable). Trivial changes to libraries or some support code can have unexpected impact on all sorts of things. 

That said, we have options here:
1. if we know what tables/datasets we want in the end, we should be able to prune the flow graph and only run tasks that are precursors to that output
2. for rapid development, the serialized intermediate results on disk can work as a cache. Because prefect can skip over tasks for which the results are already available, this boils down to explicitly _invalidating_ specific results and rerunning those tasks along with their children in the dependency graph.

File: test/unit/dfc_test.py (hunk 7 above)
Comment (rousik) [reply test/unit/dfc_test.py:128]: Yup, `get_table_names` uses `sort` so it guarantees order (with the intention to make the results stable for easy unit testing).

File: test/unit/dfc_test.py (hunk 8 above)
Comment (rousik) [reply test/unit/dfc_test.py:144]: This is a bug. The original code unioned `dfc1` and `dfc2`.

File: test/unit/dfc_test.py (hunk 9 above)
Comment (rousik) [reply test/unit/dfc_test.py:204]: yes

File: test/unit/dfc_test.py (hunk 10 above)
Comment (rousik) [reply test/unit/dfc_test.py:206]: Fanout task simply splits DFC into list of DFC with fixed number of tables each.

Comment (rousik) [reply test/unit/dfc_test.py:206]: This could be used in situations like epacems where large number of tables come together in a single dfc and where you want to process this in parallel by splitting it into smaller chunks that can be processed in parallel.

**Comment by jdangerx** [comment]
Closing this as we chose Dagster in https://github.com/catalyst-cooperative/pudl/issues/1487 and I want to clean up our open issues a bit.
